### PR TITLE
feat: test summary

### LINKED ISSUE #394: [Feature Request] Print execution summary at the end of output
Thank you so much for maintaining awesome E2E testing tool! Scenarigo has greatly improved the development productivity of me and my team.

## Issue to be solved
- If you run a lot of scenarios and some tests fail, it can be a bit tedious to scroll through the long output to see which files failed.
- In order to know how many of the tests that were run failed, it is also necessary to scroll through the long output and count the number of failed tests.

## Solution
Print test execution summary like below at the end of output.

```bash
11 tests run: 9 passed, 2 failed, 0 skipped

Failed tests:
  - scenarios/scenarigo-1.yaml
  - scenarios/scenarigo-2.yaml
```

## Implementation
- I tried to implement this feature.(https://github.com/zoncoen/scenarigo/pull/395)
- If this feature request is acceptable, I will polish up this PR on some points listed in the description of the PR.

### PR DESCRIPTION
**Pull request by kyu08**
Resolves #394

This is an implementation of #394. 
This implementation adds output like this.

| Case | Image |
|--------|--------|
|When all tests passed| image|
|When some tests failed|image|

If this feature request is acceptable, I will polish up this PR on these points.
- Make this feature configurable.
- If `colored: true` , outputs with some colors.
- Add some unit tests.
- Add feature description to README.

Thank you.

### CODE REVIEW

**Comment by zoncoen** [comment]
Thank you for your proposal!
As you suggested, it should be configurable. Please add an option to `OutputConfig`.

https://github.com/zoncoen/scenarigo/blob/ffdb628865654047c3e26b05c14d952cbdb60263/schema/config.go#L57-L61

**Comment by kyu08** [comment]
I will add few changes to this PR.
After I finished it I will reopen this PR.
Please review these changes after that. 🙏

**Review by zoncoen — CHANGES_REQUESTED** [review]

File: schema/config.go [hunk 1]
```diff
@@ -55,9 +55,10 @@ type YTTConfig struct {
 
 // OutputConfig represents an output configuration.
 type OutputConfig struct {
-	Verbose bool         `yaml:"verbose,omitempty"`
-	Colored *bool        `yaml:"colored,omitempty"`
-	Report  ReportConfig `yaml:"report,omitempty"`
+	Verbose   bool         `yaml:"verbose,omitempty"`
+	NoSummary bool         `yaml:"noSummary,omitempty"`
```
Comment (zoncoen) [inline schema/config.go:59]: Please insert `Summary` field into the line below `Colored`.

Comment (kyu08) [reply schema/config.go]: Fixed in cdf2703b061050efce6c8f45f82e5112b1b0da07

File: cmd/scenarigo/cmd/config/default.scenarigo.yaml [hunk 2]
```diff
@@ -9,6 +9,7 @@ pluginDirectory: ./gen      # Specify the root directory of plugins.
 
 output:
   verbose: false   # Enable verbose output.
+  noSummary: false # Disable show summary.
```
[the remark this answers is not in the document]
Comment (kyu08) [reply cmd/scenarigo/cmd/config/default.scenarigo.yaml]: Fixed in cdf2703b061050efce6c8f45f82e5112b1b0da07.

File: runner.go [hunk 3]
```diff
@@ -40,6 +40,8 @@ type Runner struct {
 	rootDir         string
 	inputConfig     schema.InputConfig
 	reportConfig    schema.ReportConfig
+	testSummary     testSummary
+	w               io.Writer
```
[the remark this answers is not in the document]
Comment (kyu08) [reply runner.go]: Thank you for letting me know the difference of responsibilities of Runner and `reporter`.

This point was fixed in 8a0294b673ba5a612edd71a67f8e63468e7b5e32.

**Comment by kyu08** [comment]
## TODO
- [ ] Fix CI errors.
- [ ] Add unit tests.
- [ ] Update README
    - [ ] Update code snippet generated by `scenarigo init` in `## Setup`

**Comment by kyu08** [comment]
@zoncoen
I'm done with TODOs and the changes you pointed.
This is a summary of changes after your first review.

- implement `PrintSummary` as `reporter`'s method(ce5ca6c)
- fix existing tests's errors caused by my changes (1fa298c, 5e91e6d)
- update readme (bb761bb)
- change `testSummary#add`'s argument(`r reporter.Reporter` to `testResultString string`) for the testability (6d2ac7f)
- add unit tests (a3713ee, b9375b0, f2d457b)
- format(5f31d8e)

Could you review again please?

**Review by zoncoen — CHANGES_REQUESTED** [review]
Please implement the logic to print a summary in the `reporter` package (not `scenarigo` package). I think it is enough like the following code.

```func run(cmd *cobra.Command, args []string) error {
	// ...
	if cfg.Output.Summary {
		reporterOpts = append(reporterOpts, reporter.WithTestSummary())
	}

	var reportErr error
	success := reporter.Run(
		func(rptr reporter.Reporter) {
			r.Run(context.New(rptr))
			reportErr = r.CreateTestReport(rptr)
		},
		reporterOpts...,
	)
	// ...
}```

https://github.com/zoncoen/scenarigo/blob/3f72845f973421774eecda4a60f222482e8ae5e2/cmd/scenarigo/cmd/run.go#L34

```func (r *reporter) runWithRetry(name string, f func(t Reporter), policy RetryPolicy) bool {
	// ...
	if r.isRoot() {
		printReport(child)  // This is 
		if r.context.summary {
			printSummary()
		}
	}
	return !child.Failed()
}```

https://github.com/zoncoen/scenarigo/blob/3f72845f973421774eecda4a60f222482e8ae5e2/reporter/reporter.go#L265

**Comment by kyu08** [comment]
Thank you for your feed back!

> Please implement the logic to print a summary in the reporter package (not scenarigo package). I think it is enough like the following code.

I'm sorry for misunderstood your advice. I'll fix it.

Also, thank you for showing me code examples kindly.
I tried to print `test message` in a way like your second code snippet like this.
```go
func (r *reporter) runWithRetry(name string, f func(t Reporter), policy RetryPolicy) bool {
        // ...
	if r.isRoot() {
		printReport(child)
		r.context.printf("%s\n", "\ntest message\n")
	}
	return !child.Failed()
}
```

But it seems like `test message` was showed after every test file execution like below.
image

How about call `printSummary` [here](https://github.com/kyu08/scenarigo/blob/f2d457bb2a3f96916ca463c468e7dc29ab52fcfd/reporter/reporter.go#L56) and  call `printSummary#add` [here](https://github.com/kyu08/scenarigo/blob/f2d457bb2a3f96916ca463c468e7dc29ab52fcfd/reporter/reporter.go#L284)?

When I change to call `r.context.printf` in `Run` func like below.
```go
// reporter/reporter.go
func Run(f func(r Reporter), opts ...Option) bool {
	r := run(f, opts...)
      	r.context.printf("test message called in Run func\n")
	return !r.Failed()
}
```
Then `test message` was shown just once properly.
image

**Comment by zoncoen** [comment]
Ah~ you're right. It's OK to print a summary in `Run` func 👍

**Comment by kyu08** [comment]
I fixed to  implement test summary logic in the `reporter` package.
And I added some minor changes. Please check it out.

### About the naming `summary` versus `testSummary` c138756506be77d13e8650f6991937d8e410952c
I used the naming `summary` in the context of configuration, and I used the naming `testSummary` in other contexts.
Because I feel like it is natural that the naming `summary` means this test summary feature in the config context, on the other hand it seems unnatural in the context of whole code base.

If you have any thought, please give me an idea🙏

File: reporter/reporter.go [hunk 4]
```diff
@@ -276,6 +288,7 @@ func (r *reporter) runWithRetry(name string, f func(t Reporter), policy RetryPol
 	r.appendChildren(child)
 	if r.isRoot() {
 		printReport(child)
+		child.context.testSummary.append(name, TestResultString(child))
```
Comment (zoncoen) [inline reporter/reporter.go:291]: I don't want to keep any data if the summary feature is disabled. How about like below?

```go
func (s *testSummary) append(testFileRelPath string, r Reporter) {
	if s == nil {
		return // disalbed
	}
	testResultString := TestResultString(r)
	// ...
}
```

File: reporter/test_summary.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	switch testResultString {
+	case TestResultPassed.String():
+		s.passed = append(s.passed, testFileRelPath)
+	case TestResultFailed.String():
+		s.failed = append(s.failed, testFileRelPath)
+	case TestResultSkipped.String():
+		s.skipped = append(s.skipped, testFileRelPath)
+	default: // Do nothing
+	}
+}
```
Comment (zoncoen) [inline reporter/test_summary.go:41]: It is sufficient to count only for passed and skipped. Filenames are needed only for failed.

**Review by zoncoen — CHANGES_REQUESTED** [review]

File: reporter/reporter.go [hunk 6]
```diff
@@ -47,12 +47,17 @@ type Reporter interface {
...
+	// for test summary
+	printTestSummary()
 }
 
 // Run runs f with new Reporter which applied opts.
 // It reports whether f succeeded.
 func Run(f func(r Reporter), opts ...Option) bool {
 	r := run(f, opts...)
+	r.printTestSummary()
+
```
[the remark this answers is not in the document]
Comment (kyu08) [reply reporter/reporter.go]: 🙏 
Fixed in 151da5ad3fdf94747b602d83616cbc09d5961fe6.

File: reporter/context.go [hunk 7]
```diff
@@ -38,6 +38,14 @@ func WithNoColor() Option {
 	}
 }
 
+// WithTestSummary returns an option to enable test summary.
+func WithTestSummary(enabledColor bool) Option {
+	return func(ctx *testContext) {
+		ctx.enabledTestSummary = true
+		ctx.testSummary = newTestSummary(enabledColor)
+	}
+}
```
[the remark this answers is not in the document]
Comment (zoncoen) [reply reporter/context.go:47]: > I understand you mean testSummary struct should not have enabledColor field and testContext.noColor's value should be passed to testSummary#String as needed.

That's right. In other words, `reporter. WithNoColor` option should control all output.

> So I feel there is no point to change about this.

The new implementation is also acceptable. 👍

File: reporter/reporter.go (hunk 4 above)
Comment (kyu08) [reply reporter/reporter.go]: Fixed in aca2ad33f985c60f2d3daa30938fd1ebe9e42bf4.

File: reporter/test_summary.go (hunk 5 above)
Comment (kyu08) [reply reporter/test_summary.go:42]: Fixed in 8d95a95e1a3b64b662f2a119d3e9e0aa703313f8.

**Review by zoncoen — APPROVED** [review]
Thank you for taking the time and patience to bring this into the final shape!
LGTM 💯

**Comment by kyu08** [comment]
Thank you for reviewing my clumsy code using your precious time too!🙏
And I'm sorry that lack of understanding of this project of me leads increasing time and labor to review this PR.
I probably can deal with it more smoothly next time, so I am willing to contribute if there is an issue I can solve.

**Comment by kyu08** [comment]
Thank you for reviewing my pull request the other day. I'm looking forward to using the test summary feature, and I’d really appreciate it if you could roll out the next version when it’s feasible. 
Thank you.

**Pull request merged**